### ISSUE TITLE
Performance issue on module variable expansion

### ISSUE THREAD

**Issue by horjulf** [hashicorp/terraform#28170]
Hi,

Expanding module variables has a very high performance cost as variable number increases and is limited by single core CPU performance.

500 modules, 1 string variable
`terraform plan  1.18s user 0.34s system 211% cpu 0.724 total`

500 modules, 60 string variables (Single thread 100% CPU usage)
`terraform plan  92.07s user 10.29s system 144% cpu 1:10.81 total`

### Current Terraform Version

```
Terraform v0.14.8
```

### Steps to replicate

File structure:
```
├── module
│   └── variables.tf
└── module.tf
```

1. Generate module variables
```
mkdir module
for i in {1..60}; do echo -e "variable \"var${i}\" {\n  type = string\n  default = \"\"\n}";done >! module/variables.tf
```
2. Generate module resources 
```
for i in {1..500}; do echo -e "module \"var${i}\" {\n  source = \"./module\"\n}";done >! module.tf
```
3. `terraform init`
4. `time terraform plan`

### Attempted Solutions

Reducing variable numbers drastically improves performance.

### Debug

Most expensive call seems to be
```
2021/03/23 18:18:25 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
```

**Comment by jbardin** [comment]
Hi @horjulf,

Thanks for filing the issue. Can you show an example of the configuration using these variables? 

The transitive reduction is known to be quite time consuming with large, highly connected graphs, as the number of edges can increase exponentially. Configurations of this size are not something we're optimizing for right now, but understanding the configuration may help us with incremental improvements or work arounds in the near-term. 

Thanks!

**Comment by horjulf** [comment]
Hi @jbardin,

Thanks for taking a look, the config is very basic without any dependencies, it can be generated with the steps I described.
Posting here the overview:

module.tf
```hcl
module "var1" {
  source = "./module"
}
# ...
```

module/variables.tf
```hcl
variable "var1" {
  type = string
  default = ""
}
# ...
```

It affects validate, plan and apply commands.
No other resources in the code, this escalates exponentially when there is dependencies with other modules or other complex logic (for_each, counts, etc). I tried to keep the example as simple as possible as I've noticed that this is directly linked to the number of variables in modules.

Complexity of variables doesn't seem to have any influence. e.g. object with 50 fields is the same as a string.

What I did to workaround this, was to merge multiple variables into a single object variable, this reduced `terraform validate` time from 10 to 3 minutes in a single repo by reducing the number of module variables in a state file with ~5k resources.

**Comment by jbardin** [comment]
Thanks @horjulf, I missed that the shell commands were the _entirety_ of the configuration.

While the internal graph for this is fairly large (~30k nodes, ~90k edges), there's nothing excessive or surprising there. It does however look like I did miss an optimization in profiling this code for #23811, which coincidentally was just discovered by another user in #28183. I tested with that change and the execution time drops by about 75%.
